### PR TITLE
Do not add unknown enum parameters to policy Array

### LINKED ISSUE #1921: Processing invalid PT after cutting off unknown_parameter or unknown_RPC
**Occurrence: Always** 

## Steps to reproduce
1. Make sure SDL is built with default flags
2. Start SDL and HMI
3. Connect mobile device
4. Register new application
5. Create PT for policy table update with unknown_parameter or unknown_RPC
6. Perform policy table update from mobile app with created PT

## Expected result
1. SDL received UpdatedPT with at least one <unknown_parameter> or <unknown_RPC>
and after cutting off <unknown_parameter> or <unknown_RPC> UpdatedPT performs validation of received PT
2. SDL must log the error internally and discard Policy Table Update

## Actual result
SDL performs PT validation without cutting off <unknown_parameter> or <unknown_RPC>

## Test script
[1921_Invalid_PT_after_cutting_unknow_values.lua](https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/develop/test_scripts/Defects/4_5/1921_Invalid_PT_after_cutting_unknow_values.lua)

## Attachments

# Expected delivery
- [x] Source code updates
- [x] Code comments
- [x] **UTs add/update _(not required)_**
- [x] ATF tests add/update
- [x] **Manual tests _(not required)_**
- [x] **Add/update CI plans/jobs _(not required)_**
- [x] **SDD updates _(not required)_**
- [x] **Guidelines update ([sdl_core_guides](https://github.com/smartdevicelink/sdl_core_guides)) _(not required)_**
- [x] **Guidelines update ([sdl_hmi_integration_guidelines](https://github.com/smartdevicelink/sdl_hmi_integration_guidelines)) _(not required)_**

### PR DESCRIPTION
**Pull request by LuxoftAKutsan**
Fixes #1921

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Tested with Automated test framework

### Summary

In case of unknown or invalid value of RPC parameters comes
with PTU validation must not fail but cut
off such values and continue validation of other parts of PTU.

Note that this changes cut off just PRC parameters, but do not cut off RPCs, as mentioned in #1921.
This is done because of aligning code with Sync3.2v2. Functionality of cutting of RPCs from PTU is absent in Sync3.2v2 repo. 

### CLA
- [x ] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/rpc_base/include/rpc_base/rpc_base.h [hunk 1]
```diff
@@ -307,6 +307,40 @@ class Array : public std::vector<T>, public CompositeType {
...
+// validation rules
+
+template <typename E, size_t minsize, size_t maxsize>
+class Array<Enum<E>, minsize, maxsize> : public std::vector<Enum<E> >,
+                                         public CompositeType {
+ public:
+  // Types
+  typedef std::vector<Enum<E> > ArrayType;
+
+ public:
```
Comment (AKalinich-Luxoft) [inline src/components/rpc_base/include/rpc_base/rpc_base.h]: @LuxoftAKutsan redundant public modifier

File: src/components/rpc_base/include/rpc_base/rpc_base_json_inl.h [hunk 2]
```diff
@@ -287,6 +287,61 @@ Json::Value Array<T, minsize, maxsize>::ToJsonValue() const {
...
 }
 
+////////// Array to be instantiated with Enum<T>
+
+// Non-const version
+template <typename E, size_t minsize, size_t maxsize>
+Array<Enum<E>, minsize, maxsize>::Array(Json::Value* value)
+    : CompositeType(InitHelper(value, &Json::Value::isArray)) {
+  if (value) {
+    if (value->isArray()) {
```
Comment (AKalinich-Luxoft) [inline src/components/rpc_base/include/rpc_base/rpc_base_json_inl.h]: @LuxoftAKutsan I believe these two conditions could be merged to one using &&

File: src/components/rpc_base/include/rpc_base/rpc_base_json_inl.h [hunk 3]
```diff
@@ -287,6 +287,61 @@ Json::Value Array<T, minsize, maxsize>::ToJsonValue() const {
...
+      // depending on InitHelper result
+    }
+  }
+}
+
+// Const version, must be identical to the non-const version
+template <typename E, size_t minsize, size_t maxsize>
+Array<Enum<E>, minsize, maxsize>::Array(const Json::Value* value)
+    : CompositeType(InitHelper(value, &Json::Value::isArray)) {
+  if (value) {
```
Comment (AKalinich-Luxoft) [inline src/components/rpc_base/include/rpc_base/rpc_base_json_inl.h]: @LuxoftAKutsan is it possible to use const version of ctor in non-const ctor instead of code duplication? Especially if they must be identical

File: src/components/policy/policy_regular/include/policy/policy_table/generated_MOBILE_API_policy_types.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
+/**
+ * @file generated_MOBILE_API_policy_types.h
+ * @brief Generated class generated_MOBILE_API_policy_types header file.
+ *
+* Copyright (c) 2017, Ford Motor Company
```
Comment (AByzhynar) [inline src/components/policy/policy_regular/include/policy/policy_table/generated_MOBILE_API_policy_types.h]: @LuxoftAKutsan Change year in files  generation

File: src/components/policy/policy_regular/include/policy/policy_table/generated_msg_version.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
+/*Copyright (c) 2016, Ford Motor Company
```
Comment (AByzhynar) [inline src/components/policy/policy_regular/include/policy/policy_table/generated_msg_version.h]: @LuxoftAKutsan Change year in files  generation

File: src/components/policy/policy_regular/include/policy/policy_table/generated_msg_version.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+#ifndef GENERATED_MSG_VERSION_H
+#define GENERATED_MSG_VERSION_H
+
+namespace application_manager {
+
+const uint16_t major_version = 4;
+const uint16_t minor_version = 5;
+const uint16_t patch_version = 0;
+}  // namespace application_manager
+#endif  // GENERATED_MSG_VERSION_H
```
Comment (AByzhynar) [inline src/components/policy/policy_regular/include/policy/policy_table/generated_msg_version.h]: @LuxoftAKutsan Is it possible to enable empty string adding at the end of generated files

File: src/components/application_manager/src/application_manager_impl.cc [hunk 7]
```diff
@@ -2853,14 +2853,21 @@ void ApplicationManagerImpl::UnregisterAllApplications() {
 
   hmi_cooperating_ = false;
   bool is_ignition_off = false;
-  using namespace mobile_api::AppInterfaceUnregisteredReason;
```
Comment (yang1070) [inline src/components/application_manager/src/application_manager_impl.cc]: the new code is equivalent to the old code, what's the reason behind this change (not using name space in this file)?

File: src/components/policy/policy_regular/include/policy/policy_table/enums.h [hunk 8]
```diff
@@ -93,23 +94,23 @@ bool IsValidEnum(Parameter val);
...
-  AHT_BACKGROUND_PROCESS,
-  AHT_TESTING,
-  AHT_SYSTEM,
-  AHT_PROJECTION,
-  AHT_REMOTE_CONTROL
-};
-bool IsValidEnum(AppHMIType val);
-const char* EnumToJsonString(AppHMIType val);
-bool EnumFromJsonString(const std::string& literal, AppHMIType* result);
+// enum AppHMIType {
```
Comment (yang1070) [inline src/components/policy/policy_regular/include/policy/policy_table/enums.h]: why commented them out and left them in code, why not just delete them

File: src/components/policy/policy_regular/src/policy_table/enums.cc [hunk 9]
```diff
@@ -302,107 +302,107 @@ bool EnumFromJsonString(const std::string& literal, Parameter* result) {
...
-    *result = AHT_PROJECTION;
-    return true;
-  } else if ("REMOTE_CONTROL" == literal) {
-    *result = AHT_REMOTE_CONTROL;
-    return true;
-  } else {
-    return false;
-  }
-}
+// bool IsValidEnum(AppHMIType val) {
```
Comment (yang1070) [inline src/components/policy/policy_regular/src/policy_table/enums.cc]: why keep these commented code?

File: src/components/application_manager/src/application_manager_impl.cc (hunk 7 above)
Comment (AByzhynar) [reply src/components/application_manager/src/application_manager_impl.cc]: @yang1070 You are right. The code is equal to the old one regarding functionality. But to avoid namespace pollution and to improve readability of code clear usage of entities form namespace were added. 
`mobile_api::AppInterfaceUnregisteredReason::INVALID_ENUM` is more clear than just `INVALID_ENUM`, isn't it?

File: src/components/policy/policy_regular/include/policy/policy_table/enums.h (hunk 8 above)
Comment (AByzhynar) [reply src/components/policy/policy_regular/include/policy/policy_table/enums.h]: @yang1070 Fixed in d1308970ca2afad7de251f6ef744d94c4422cb42

File: src/components/policy/policy_regular/src/policy_table/enums.cc (hunk 9 above)
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/policy_table/enums.cc]: @yang1070 Fixed in d1308970ca2afad7de251f6ef744d94c4422cb42

File: src/components/rpc_base/include/rpc_base/rpc_base.h (hunk 1 above)
Comment (AByzhynar) [reply src/components/rpc_base/include/rpc_base/rpc_base.h]: @AKalinich-Luxoft Fixed in d1308970ca2afad7de251f6ef744d94c4422cb42

File: src/components/rpc_base/include/rpc_base/rpc_base_json_inl.h (hunk 2 above)
Comment (AByzhynar) [reply src/components/rpc_base/include/rpc_base/rpc_base_json_inl.h]: @AKalinich-Luxoft Fixed in d1308970ca2afad7de251f6ef744d94c4422cb42

File: src/components/rpc_base/include/rpc_base/rpc_base_json_inl.h (hunk 3 above)
Comment (AByzhynar) [reply src/components/rpc_base/include/rpc_base/rpc_base_json_inl.h]: @AKalinich-Luxoft Fixed in d1308970ca2afad7de251f6ef744d94c4422cb42

**Comment by AByzhynar** [comment]
@yang1070 Do you have more comments or the changes are good for you?

**Comment by AByzhynar** [comment]
Rebuild required

**Comment by mrapitis** [comment]
@JackLivio @jacobkeeler can you please review when time permits? thanks!

**Comment by jacobkeeler** [comment]
Duplicate of #2006
